SymCryptProvider/src/p_scossl_rand.c

```
@@ -0,0 +1,137 @@
+//
+// Copyright (c) Microsoft Corporation. Licensed under the MIT license.
+//
+
+#include <openssl/core_dispatch.h>
+#include <openssl/core_names.h>
+#include <openssl/proverr.h>
+
+#include "scossl_helpers.h"
+
+#define SYMCRYPT_DRGB_STRENGTH 256
+#define SYMCRYPT_DRGB_MAX_REQUEST_SIZE (1 << 16)
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+const BYTE dummy_scossl_ctx = 0;
+
+static const OSSL_PARAM gettable_ctx_param_types[] = {
+    OSSL_PARAM_int(OSSL_RAND_PARAM_STATE, NULL),
+    OSSL_PARAM_uint(OSSL_RAND_PARAM_STRENGTH, NULL),
+    OSSL_PARAM_size_t(OSSL_RAND_PARAM_MAX_REQUEST, NULL),
+    OSSL_PARAM_END};
+
+// All state is maintained inside SymCrypt, but newctx and
+// freectx are expected. These do nothing in the SymCrypt provider
+static void *p_scossl_rand_newctx(ossl_unused void *provctx,
+                                  ossl_unused void *parent,
+                                  ossl_unused const OSSL_DISPATCH *parent_calls)
+{
+    return (void *)&dummy_scossl_ctx;
+}
+
+static void p_scossl_rand_freectx(ossl_unused void *ctx){}
+
+// RNG state is internally managed by SymCrypt. This function is
+// required, but does not actually instantiate SymCrypt's RNG state
+static SCOSSL_STATUS p_scossl_rand_instantiate(ossl_unused void *ctx,
+                                               ossl_unused unsigned int strength,
+                                               ossl_unused int prediction_resistance,
+                                               _In_reads_bytes_opt_(addin_len) const unsigned char *addin, size_t addin_len,
+                                               ossl_unused const OSSL_PARAM params[])
+{
```

Maybe this should return an error if the requested strength > 256 or prediction resistance is non-zero?

---

There's no defined expected behavior but that makes sense to me.

---

I think you got the check exactly backwards - we should fail if the caller requests prediction resistance (not supported) or asks for a strength which is higher than we support.